### PR TITLE
Adds support for Elasticsearch 7.x

### PR DESCRIPTION
**Pull request by codefromthecrypt**
There are multiple issues to resolve, not just the colon banning.

See https://www.elastic.co/guide/en/elasticsearch/reference/7.x/breaking-changes-7.0.html
See #2219

### CODE REVIEW

**Comment by codefromthecrypt** [comment]
the thing breaks on lack of default mapping. We have some things there, so we have to be careful how to address this. I'm putting this change down because I didn't anticipate this being more than just the colon problem.

**Comment by making** [comment]
Elasticsearch 7.0 has been released. 
Is there any workaround to change the index pattern zipkin uses?

**Comment by codefromthecrypt** [comment]
@making it is not just an issue of the index naming convention. that's a very small matter actually. they've made significant changes to the index template itself

**Comment by making** [comment]
@adriancole so isn't ES 7.0 available until this issue is fixed?

**Comment by codefromthecrypt** [comment]
@making this pull request needs to pass :) the index template needs to work and it doesn't on ES 7, so that's the main issue. When done debugging other things, I'll throw some time on this. It is good to know that ES 7 is final now.

**Comment by codefromthecrypt** [comment]
there's actually another significant issue which is the revlock on the dependency aggregation job. To have dependencies link on ES 7 probably implies breaking ES 2.x https://github.com/openzipkin/zipkin-dependencies/issues/124 cc @openzipkin/elasticsearch

**Comment by making** [comment]
Do you continue to support ES 2x?

**Comment by codefromthecrypt** [comment]
@making we've not yet unsupported it, so yeah we run tests on it on every build. I don't think we would want to un-support it at least until lens is default as otherwise people will revlock into the old UI.

amazon still lets people spin up 1.5! https://aws.amazon.com/elasticsearch-service/faqs/

**Comment by codefromthecrypt** [comment]
I found the two blockers on zipkin-dependencies. holding until they are sorted as it will not make sense to proceed here if we can't aggregate dependencies.

https://github.com/elastic/elasticsearch-hadoop/issues/1277
https://github.com/elastic/elasticsearch-hadoop/issues/1276

**Comment by codefromthecrypt** [comment]
@gquintana until tests pass there's no point in doing work that feigns ES 7 support. main thing needed here is to finish the drift. there was a dramatic change in index templates introduced in ES 7, iotw changing the pattern doesn't actually make things work. this PR starts the process and if you want you can finish this change if I don't finish it first.

right now, we are having a workshop, so I won't be working on this for the next few days. separately ES have no working version of elasticsearch-hadoop anyway as noted by issues not yet acked by ES folks
https://github.com/elastic/elasticsearch-hadoop/issues/1276
https://github.com/elastic/elasticsearch-hadoop/issues/1277

File: zipkin-storage/elasticsearch/src/main/java/zipkin2/elasticsearch/ElasticsearchSpanConsumer.java [hunk 1]
```diff
@@ -47,19 +48,26 @@
...
   final IndexNameFormatter indexNameFormatter;
+  final char indexTypeDelimiter;
   final boolean searchEnabled;
   final DelayLimiter<AutocompleteContext> delayLimiter;
 
   ElasticsearchSpanConsumer(ElasticsearchStorage es) {
     this.es = es;
     this.autocompleteKeys = new LinkedHashSet<>(es.autocompleteKeys());
     this.indexNameFormatter = es.indexNameFormatter();
+    this.indexTypeDelimiter = es.indexTypeDelimiter();
```
Comment (codefromthecrypt) [inline zipkin-storage/elasticsearch/src/main/java/zipkin2/elasticsearch/ElasticsearchSpanConsumer.java:62]: super note: literally this is not enough! let's please defocus from this being the single issue blocking ES 7 support

**Comment by codefromthecrypt** [comment]
when this change is done, nagging on this issue will be welcome again :) https://github.com/apache/incubator-zipkin/issues/2513

hopefully, that's not longer than a week.

**Comment by codefromthecrypt** [comment]
current status. no longer any errors installing the index template on ES 7. However, all the query tests are broke

**Comment by codefromthecrypt** [comment]
a couple more mapping problems..

```
java.lang.IllegalStateException: {"took":676,"errors":true,"items":[{"index":{"_index":"gettagsandvalues-span-2019-05-03","_type":"span","_id":"rQgAfGoBmB-cc_Lhh4f3","status":500,"error":{"type":"mapper_exception","reason":"Can't update attribute for type [span.tags.enabled] in index mapping"}}},{"index":{"_index":"gettagsandvalues-autocomplete-2019-05-03","_type":"autocomplete","_id":"http.host=host1","status":400,"error":{"type":"illegal_argument_exception","reason":"Rejecting mapping update to [gettagsandvalues-autocomplete-2019-05-03] as the final mapping would have more than 1 type: [_doc, autocomplete]"}}}]}
```

**Comment by codefromthecrypt** [comment]
@making I tested this branch and it now works (eventhough the spark job is still messed up for reasons noted multiple times). Can you or someone else verify?

```bash
$ TAG=elasticsearch7-SNAPSHOT
$ curl -sSL https://jitpack.io/com/github/apache/incubator-zipkin/zipkin-server/${TAG}/zipkin-server-${TAG}-exec.jar > zipkin.jar
$ STORAGE_TYPE=elasticsearch java -jar zipkin.jar 
```

**Comment by making** [comment]
@adriancole Thanks! I'll try it on my env.

**Comment by codefromthecrypt** [comment]
actually this week is still golden week in Japan. @gquintana care to give a swing? ^^ for a temporary download location

**Comment by codefromthecrypt** [comment]
hah wires crossed @making thanks for not being on holiday :)

**Comment by making** [comment]
No problem, it's a golden development week :)

I confirmed that Zipkin started to create a `zipkin-span-{yyyy}-{mm}-{dd}` index.

image

But it's weird to see only one service sends traces ... (There should be 3 or 4 services.)

image

image

**Comment by codefromthecrypt** [comment]
Thanks for verifying, @making the service change could be unrelated recent change landed in the next pending release https://github.com/apache/incubator-zipkin/releases/tag/v2.13.0

**Comment by codefromthecrypt** [comment]
I'll merge this. this commit doesn't remove support for ES 2.x or anything just adds 7.x. More follow-ups will happen mostly in https://github.com/openzipkin/zipkin-dependencies/ and this change will go out in the next server release after 2.13 (likely will be named 2.14)

**Comment by codefromthecrypt** [comment]
the above jitpack curl command should also work for master now (`TAG=master-SNAPSHOT`)

**Comment by codefromthecrypt** [comment]
ps I just tried with brave-webmvc-example just to verify service names are ok. I think it is ok.

Screen Shot 2019-05-03 at 7 19 10 PM

**Comment by chefky** [comment]
Trying the latest snapshot with elasticsearch 7.0 -- getting an issue when
I update the zipkin-span_template. I need the tags field to be searchable
since I am using tags.http.status_code to determine success rate of an API.
I just got the default value that is created by zipkin via 'GET
/_template/zipkin-span_template' in kibana, and edited the tags field to:

[image: Screen Shot 2019-05-06 at 4.03.24 PM.png]
Got this to work prior to latest snapshot + elasticsearch 7.0, but now I
get an error:

2019-05-06 15:47:55.847 DEBUG 21101 --- [und.io:9243/...]
z.s.i.ZipkinHttpCollector                : Cannot store spans
[327a37f9ae4eaa6d/52a9ba8e5a3f60f9, 327a37f9ae4eaa6d/87f3b8e489f2895b] due
to
IllegalStateException({"took":801,"errors":true,"items":[{"index":{"_index":"zipkin-span-2019-05-06","_type":"span","_id":"WewdjGoBqE8vRr9QDFH6","status":400,"error":{"type":"illegal_argument_exception","reason":"Can't
merge because of conflicts: [Cannot update excludes setting for
[_source]]"}}},{"index":{"_index":"zipkin-span-2019-05-06","_type":"span","_id":"WuwdjGoBqE8vRr9QDFH6","status":400,"error":{"type":"illegal_argument_exception","reason":"Can't
merge because of conflicts: [Cannot update excludes setting for
[_source]]"}}}]})


java.lang.IllegalStateException:
{"took":801,"errors":true,"items":[{"index":{"_index":"zipkin-span-2019-05-06","_type":"span","_id":"WewdjGoBqE8vRr9QDFH6","status":400,"error":{"type":"illegal_argument_exception","reason":"Can't
merge because of conflicts: [Cannot update excludes setting for
[_source]]"}}},{"index":{"_index":"zipkin-span-2019-05-06","_type":"span","_id":"WuwdjGoBqE8vRr9QDFH6","status":400,"error":{"type":"illegal_argument_exception","reason":"Can't
merge because of conflicts: [Cannot update excludes setting for
[_source]]"}}}]}

at
zipkin2.elasticsearch.internal.HttpBulkIndexer$CheckForErrors.convert(HttpBulkIndexer.java:60)
~[zipkin-storage-elasticsearch-2.13.1-SNAPSHOT.jar!/:2.13.1-SNAPSHOT]

at
zipkin2.elasticsearch.internal.HttpBulkIndexer$CheckForErrors.convert(HttpBulkIndexer.java:53)
~[zipkin-storage-elasticsearch-2.13.1-SNAPSHOT.jar!/:2.13.1-SNAPSHOT]

at
zipkin2.elasticsearch.internal.client.HttpCall.parseResponse(HttpCall.java:151)
~[zipkin-storage-elasticsearch-2.13.1-SNAPSHOT.jar!/:2.13.1-SNAPSHOT]

at
zipkin2.elasticsearch.internal.client.HttpCall$V2CallbackAdapter.onResponse(HttpCall.java:128)
[zipkin-storage-elasticsearch-2.13.1-SNAPSHOT.jar!/:2.13.1-SNAPSHOT]

at okhttp3.RealCall$AsyncCall.execute(RealCall.java:174)
[okhttp-3.14.0.jar!/:?]

at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
[okhttp-3.14.0.jar!/:?]

at
java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[?:1.8.0_191]

at
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[?:1.8.0_191]
at java.lang.Thread.run(Thread.java:748) [?:1.8.0_191]

Tried to remove setting for _source but then I get this:

"Rejecting mapping update to [zipkin-span-2019-05-06] as the final mapping
would have more than 1 type: [_doc, span]"

Any ideas would be appreciated.

Thanks,
Karen

**Comment by codefromthecrypt** [comment]
are you saying you have a secondary index template and are trying to merge
this with the one we create?

aside: conventionally we add the tag error when http status is error

**Comment by chefky** [comment]
Am trying to overwrite the one that you created, but I do this in Kibana,
after the default template has been created. Original tags field has
enabled = false, so I just run PUT _template/zipkin-span_template with
everything the same as default, except for the tags field. This is saved
successfully, but the next time I run my test program and new spans are
created, I get the merge conflict. Since this was working in previous
versions, I thought zipkin checked for existing templates and did nothing
if one was found.

On the side note : I need to track specific error codes as we have some
exceptions (e.g. 4XX codes, as well as 503, do not affect success rate)

**Comment by codefromthecrypt** [comment]
Custom index templates are not covered in the code here. Let's move to chat
and help troubleshoot to the degree we can?
https://gitter.im/openzipkin/zipkin

**Comment by codefromthecrypt** [comment]
@chefky it is near impossible for us to warranty anything custom with elasticsearch especially as versions change. keeping that in mind, I spent an hour trying to figure out the magic incantation and it "seems to work" on my laptop. If you need more support on elasticsearch, probably best to ask them as they know the nuance here. anyway with all disclosures aside.. https://gist.github.com/adriancole/1af1259102e7a2da1b3c9103565165d7

**Comment by pete-woods** [comment]
Thanks for fixing this so quickly, folks. Looking forward to `2.14`!

**Comment by making** [comment]
my update:

> But it's weird to see only one service sends traces ... (There should be 3 or 4 services.)

All services needed to be restarted to get all traces again. It works fine now :)

**Comment by nopeno** [comment]
when i installed elasticsearch 7.2.0, i got:

failed to put template [zipkin:span_template]
java.lang.IllegalArgumentException: Setting index.mapper.dynamic was removed after version 6.0.0
        at org.elasticsearch.index.mapper.MapperService.<init>(MapperService.java:161) ~[elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.index.IndexService.<init>(IndexService.java:175) ~[elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.index.IndexModule.newIndexService(IndexModule.java:399) ~[elasticsearch-7.2.0.jar:7.

**Comment by shakuzen** [comment]
@nopeno it would be better to chat on Gitter, since this is an already merged pull request. Please make sure you're using the latest version of the Zipkin Server, and if you're still having issues, let us know at https://gitter.im/openzipkin/zipkin

**Pull request merged**